scripts/validate-schemas-final.js

```diff
@@ -0,0 +1,104 @@
+   const directoryPath = './schemas';
+
+
+   const files = fs.readdirSync(directoryPath);
+
+  // Filter files based on start and end file names
+  const filteredFiles = files.filter(file => {
+    return file >= startFileName && file <= endFileName;
+  });
+
+
+    // Iterate through the filtered files
+  filteredFiles.forEach(file => {
+    // Construct the full path to the JSON schema file
+    const filePath = path.join(directoryPath, file);
+
+    try {
+      // Read and parse the JSON schema
+      const fileContent = fs.readFileSync(filePath, 'utf8');
+      const obj = JSON.parse(fileContent);
+
+      // Remove unnecessary definitions
+      delete obj.definitions['http://json-schema.org/draft-04/schema'];
+      delete obj.definitions['http://json-schema.org/draft-07/schema'];
+
+      // Compile the schema
+      const validate = ajv.validateSchema(obj);
+
+      // Check if the schema is valid
+      if (validate) {
+        console.log(`${file}: JSON Schema is valid!`);
+      } else {
+        console.error(`${file}: JSON Schema is not valid:`, ajv.errors);
+        process.exit(1);
+      }
+    } catch (error) {
+      console.error(`${file}: Error reading or parsing JSON Schema:`, error.message);
+      process.exit(1);
+    }
+  });
+}
+
+
+function validation(draft, startFileName, excludedFiles){
+
+    const Ajv = draft === 'draft04' ? require('ajv-draft-04') : require('ajv');
+    const ajv = new Ajv();
+ 
+    // Specify the path to the 'schemas' directory
+    const directoryPath = './schemas';
+ 
+ 
+    const files = fs.readdirSync(directoryPath);
+ 
+   // Filter files based on start and end file names
+   const filteredFiles = files.filter(file => {
+     return file >= startFileName && !excludedFiles.includes(file);
+   });
+ 
+ 
+    // Iterate through the filtered files
+   filteredFiles.forEach(file => {
+     // Construct the full path to the JSON schema file
+     const filePath = path.join(directoryPath, file);
+ 
+     try {
+       // Read and parse the JSON schema
+       const fileContent = fs.readFileSync(filePath, 'utf8');
+       const obj = JSON.parse(fileContent);
+ 
+       // Remove unnecessary definitions
+       delete obj.definitions['http://json-schema.org/draft-04/schema'];
+       delete obj.definitions['http://json-schema.org/draft-07/schema'];
+ 
+       // Compile the schema
+       const validate = ajv.validateSchema(obj);
+ 
+       // Check if the schema is valid
+       if (validate) {
+         console.log(`${file}: JSON Schema is valid!`);
+       } else {
+         console.error(`${file}: JSON Schema is not valid:`, ajv.errors);
+         process.exit(1);
+       }
+     } catch (error) {
+       console.error(`${file}: Error reading or parsing JSON Schema:`, error.message);
+       process.exit(1);
+     }
+   });
+}
+
+
+validationDraft04('draft04', '1.0.0-without-$id.json', '1.2.0.json');
+validation('draft07', '2.0.0-rc1.json', ['README.md', 'all.schema-store.json']);
```

Comment:
I' a bit concerned about the fact you rely on alphabetical order for picking up the schemas to validate. I think this strategy can create several problems in the future, problems that could be hidden and ending up in schemas not being validated because they fall alphabetically after, in this case, `all.schema-store.json`. 

Do you think we can find a better alternative to this?
As a matter of fact, I can tell you no more draft-4 is gonna be used besides those few files, and draft7 is gonna be used in long term so we can consider it as the default.

Reply:
> I' a bit concerned about the fact you rely on alphabetical order for picking up the schemas to validate. I think this strategy can create several problems in the future, problems that could be hidden and ending up in schemas not being validated because they fall alphabetically after, in this case, all.schema-store.json.

okay, I understand your point.

> Do you think we can find a better alternative to this?

Let me think and I'll suggest something

> As a matter of fact, I can tell you no more draft-4 is gonna be used besides those few files, and draft7 is gonna be used in long term so we can consider it as the default.

Okay, that's a bit of relief for me to hear :sweat_smile:

Reply:
>  think this strategy can create several problems in the future, problems that could be hidden and ending up in schemas not being validated because they fall alphabetically after, in this case, all.schema-store.json.

could you please provide me some cases which are supposed to be failed?

Reply:
I tried running the script for the `all.schema-store.json` and it gives the following output :point_down: 

```
all.schema-store.json: Error reading or parsing JSON Schema: Cannot convert undefined or null to object
```

Reply:
> think this strategy can create several problems in the future, problems that could be hidden and ending up in schemas not being validated because they fall alphabetically after, in this case, all.schema-store.json

I modified the script a bit for the `validation` function :point_down: 
```js
function validation (excludedFiles){

  const Ajv = require('ajv');
  const ajv = new Ajv();

  const directoryPath = './schemas';


  const files = fs.readdirSync(directoryPath);

  const filteredFiles = files.filter(file => !excludedFiles.includes(file));


  // Iterate through the filtered files
  filteredFiles.forEach(file => {
    // Construct the full path to the JSON schema file
    const filePath = path.join(directoryPath, file);

    try {
      // Read and parse the JSON schema
      const fileContent = fs.readFileSync(filePath, 'utf8');
      const obj = JSON.parse(fileContent);

      // Remove unnecessary definitions
      delete obj.definitions['http://json-schema.org/draft-04/schema'];
      delete obj.definitions['http://json-schema.org/draft-07/schema'];

      // Validate the schema
      const validate = ajv.validateSchema(obj);

      // Check if the schema is valid
      if (validate) {
        console.log(`${file}: JSON Schema is valid!`);
      } else {
        console.error(`${file}: JSON Schema is not valid:`, ajv.errors);
        process.exit(1);
      }
    } catch (error) {
      console.error(`${file}: Error reading or parsing JSON Schema:`, error.message);
      process.exit(1);
    }
  });
}


const excludeFiles=['README.md', 'all.schema-store.json', '1.0.0-without-$id.json', '1.0.0.json', '1.1.0-without-$id.json','1.1.0.json', '1.2.0-without-$id.json', '1.2.0.json', '2.0.0-rc1-without-$id.json'];

validationDraft04('draft04', '1.0.0-without-$id.json', '1.2.0.json');

validation(excludeFiles);
```
made a `excludeFiles` array containing the files to exclude which are `README.md`, `all.schema-store.json` and the draft-4 files. Passed it to the `validation` function and made a check inside the function to read all files except these files.

For the cases which I could think of as per my current level of knowledge, I created 2 files named `0.0.0.json` and `.es.json` for the sake of testing. Executed the script for these files and both these files were read during the execution.

These are the output for the updated script :point_down: 
```console
1.0.0-without-$id.json: JSON Schema is valid!
1.0.0.json: JSON Schema is valid!
1.1.0-without-$id.json: JSON Schema is valid!
1.1.0.json: JSON Schema is valid!
1.2.0-without-$id.json: JSON Schema is valid!
1.2.0.json: JSON Schema is valid!
.es.json: JSON Schema is valid!
0.0.0.json: JSON Schema is valid!
2.0.0-rc1.json: JSON Schema is valid!
2.0.0-rc2-without-$id.json: JSON Schema is valid!
2.0.0-rc2.json: JSON Schema is valid!
2.0.0-without-$id.json: JSON Schema is valid!
2.0.0.json: JSON Schema is valid!
2.1.0-without-$id.json: JSON Schema is valid!
2.1.0.json: JSON Schema is valid!
2.2.0-without-$id.json: JSON Schema is valid!
2.2.0.json: JSON Schema is valid!
2.3.0-without-$id.json: JSON Schema is valid!
2.3.0.json: JSON Schema is valid!
2.4.0-without-$id.json: JSON Schema is valid!
2.4.0.json: JSON Schema is valid!
2.5.0-without-$id.json: JSON Schema is valid!
2.5.0.json: JSON Schema is valid!
2.6.0-without-$id.json: JSON Schema is valid!
2.6.0.json: JSON Schema is valid!
```
Those two files are also being read along with the other draft-7 files. Is this correct? Or am I missing something? :eyes:

Reply:
Please read https://github.com/asyncapi/spec-json-schemas/pull/452#discussion_r1403103955

Reply:
> BTW, no file exclusion for json files under /schema should happen at a very first glance.

Got it!

Reply:
> I' a bit concerned about the fact you rely on alphabetical order for picking up the schemas to validate. I think this strategy can create several problems in the future, problems that could be hidden and ending up in schemas not being validated because they fall alphabetically after, in this case, `all.schema-store.json`.
> 
> Do you think we can find a better alternative to this? As a matter of fact, I can tell you no more draft-4 is gonna be used besides those few files, and draft7 is gonna be used in long term so we can consider it as the default.

Found an alternative for this.
Resolved with eccb4ab